### PR TITLE
Future API write-up

### PR DESCRIPTION
**Pull request by pcwinters**
First pass at writing up documentation for a proposed API that conceptually decouples Adrenaline from Redux, stores, and rendering strategies.

[README](https://github.com/gyzerok/adrenaline/tree/api_readme)

### CODE REVIEW

**Comment by pcwinters** [comment]
@gyzerok, this is Adrenaline as I see it. I'm sure I'm missing a few things that I'll uncover during implementation, but this API would move an awful lot of things around to make Adrenaline more flexible. It's probably easiest to view the [README](https://github.com/gyzerok/adrenaline/tree/api_readme) on my branch, particularly the Concepts and API sections.

File: README.md [hunk 1]
```diff
@@ -9,76 +9,194 @@ Adrenaline
...
+  * `args(props): Object`) A transformation of props that affect changes in your container's query requirements. Args are optional, because queries can be computed directly from props. But, args
+  will be spread as props along the same lifecycle as the data resolved for queries. In this way, you
+  can be sure that certain props (like id's or filters) that are tightly coupled to queries are updated
+  when the associated data is loaded.
+  * `queries(args): Object`) A function that returns a map of queries (the schema for which should be understood by your adapter strategy for fetching data) keyed by prop names.
+* Adapter - The Adrenaline adapter describes the strategy for fetching and resolving data from
+  both your API and cache. The adapter will dispatch actions when necessary to cache data with
+  a paired reducer. When implementing your own reducer, it's recommended to keep the reducer
+  close and in mind due to the tight relationship between data resolution and cache actions.
+* Container - A container represents a node in your redux tree that describes and resolves
```
Comment (gyzerok) [inline README.md]: I guess you mean here React tree, not Redux. Am I right?

**Comment by gyzerok** [comment]
This is great overview! I'll gonna find time today and try to add my own thoughts to this.

File: README.md (hunk 1 above)
Comment (pcwinters) [inline README.md]: :+1: yes. I wrote most of this in my post soccer game sleeplessness. There are probably other minor mistakes. I'll push some updates as I refine it.

File: README.md [hunk 2]
```diff
@@ -9,76 +9,194 @@ Adrenaline
...
+            })
+        }
+      }))
+      .then(() => dataProps); // resolve the data props
+  }
+}
+```
 
-### Only `id`
+### `<AdrenalineProvider getState, subscribe, dispatch>`
```
Comment (pcwinters) [inline README.md]: TODO add `adapter` to prop list

File: README.md [hunk 3]
```diff
@@ -9,76 +9,194 @@ Adrenaline
...
+      stateSelector={ (state) => state.adrenalineCache }
+      actionType={UPDATE_CACHE} >
+      { /* Your application */ }
+    </AdrenalineReduxProvider>
+  </Provider>
+  , document.body);
 ```
 
-### `id` is required
+### `createContainer(options)`
```
Comment (pcwinters) [inline README.md]: you could always adopt the same terminology as Relay here. This would be a 'rootContainer'.

**Comment by gyzerok** [comment]
@pcwinters the one thing I can tell you immediately: I think you are right about sticking to Redux right now. We don't need to decouple from state manager for 1.0. So assume we creating Adrenaline and it uses Redux as a state manager.

**Comment by pcwinters** [comment]
@gyzerok, you would be decoupled from Redux, with a nod to it as a sort of sanctioned state manager. The `AdrenalineProvider` (or the out of the box `AdrenalineReduxProvider` implementation`) defines the interface by which containers can subscribe/dispatch and select state from the store.

**Comment by gyzerok** [comment]
Okey, here is my thoughts on the topic.

I think that Adaptors you introduce is some kind of advanced concept. Most of the Adrenaline developers do not need this concept and I suppose they gonna use Adrenaline with GraphQL and with the default Adaptor. So currently the top level API for us is Container API and Provider API. To starting with the documentation we need to agree with both. In my opinion we are stucked because we are trying to move in every direction at the same time. Let's agree with the things one by one and finalise them in docs before any subsequent rewrites.

The first thing on the way is Container. I definitely agree with your thought on making decorator for it and even more I think we need remove `createContainer` completely. As decorator container would be easily composable with defferent higher-order stores from other libraries (e.x. Redux).

```js
import { adrenaline } from 'adrenaline';

function ExampleComponent({ resource1, resource2 }) {
  // implementation
}

const createContainer = compose(
  connect(mapStateToProps, mapDispathToProps),
  adrenaline({
     args: () => ({}),
     queries: {
       resource1: () => // query description,
       resource2: () => // query description
     },
  }),
);

export default createContainer(ExampleComponent);
``` 

* `adrenaline` - decorator
* `args` - optional pure function of props, if not presented we use empty object for args (I've missed it in last commits, but its necessary for GraphQL implementation)
* `queries` - `Object<String, Function>` represents queries; functions get computed args as input and returns query description; note: we do not declare `queries` as function, but every its resource as a function.

If you agree with this formula lets fix it in the docs and use as a reference.

The next step - Provider. Let me think on this till tomorrow. After some thinking I do really like your idea with
```js
<ReduxProvider>
  <AdrenalineProvider>
    <App />
  </AdrenalineProvider>
</ReduxProvider>
```
I just interested in discovering more fancier API for creating adrenaline's reducer.

**Comment by gyzerok** [comment]
Also if you are comfortable with it I suggest to completely remove all my legacy stuff from docs (like cache structure description).

**Comment by gyzerok** [comment]
Also `Container` should expose `getSpecs` static method which simply returns specs for testing purposes.

**Pull request closed without merging**